### PR TITLE
pinentry: 1.1.0 -> 2020-07-17

### PR DESCRIPTION
**Pull request by bagnaram**
(Support for pinentry-efl. Pinentry 1.1.0 relase from 2017, without a
recent tagged release)

###### Motivation for this change
Pinentry hasn't seen an update to its release since 2017,  however there has been quite a bit of development, including an addition of pinentry-efl for Elementary toolkit DEs like enlightenment. This change rebuilds pinentry from MASTER and includes `pinentry-efl`

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bagnaram** [comment]
I would like to include some logic in this package so that when ${version} = "master" then it builds from source, otherwise use the pre-defined packages. Any suggestions on this approach?

File: pkgs/tools/security/pinentry/default.nix [hunk 1]
```diff
@@ -35,29 +36,29 @@ let
...
 
 pinentryMkDerivation rec {
   pname = "pinentry";
-  version = "1.1.0";
+  version = "ae584040d14a4d7d6eed8be77844126130ef0118";
 
-  src = fetchurl {
-    url = "mirror://gnupg/pinentry/${pname}-${version}.tar.bz2";
-    sha256 = "0w35ypl960pczg5kp6km3dyr000m1hf0vpwwlh72jjkjza36c1v8";
+  src = builtins.fetchGit {
```
Comment (jtojnar) [inline pkgs/tools/security/pinentry/default.nix:48]: At the moment, fixed-output derivations are still required so you need to use `fetchgit` with `sha256`.

File: pkgs/tools/security/pinentry/default.nix [hunk 2]
```diff
@@ -35,29 +36,29 @@ let
...
     emacs = { bin = "emacs"; flag = "emacs"; buildInputs = []; };
+    efl = { bin = "efl"; flag = "efl"; buildInputs = [ pkgs.enlightenment.efl ]; };
   };
 
 in
 
 pinentryMkDerivation rec {
   pname = "pinentry";
-  version = "1.1.0";
+  version = "ae584040d14a4d7d6eed8be77844126130ef0118";
```
Comment (jtojnar) [inline pkgs/tools/security/pinentry/default.nix:46]: According to https://nixos.org/nixpkgs/manual/#sec-package-naming, the `version` should be `unstable-yyyy-mm-dd`, when using git snapshots.

File: pkgs/tools/security/pinentry/default.nix [hunk 3]
```diff
@@ -35,29 +36,29 @@ let
...
-    sha256 = "0w35ypl960pczg5kp6km3dyr000m1hf0vpwwlh72jjkjza36c1v8";
+  src = builtins.fetchGit {
+    url = "https://github.com/gpg/pinentry.git";
+    ref = "${version}";
   };
 
   nativeBuildInputs = [ pkgconfig autoreconfHook ]
     ++ concatMap(f: flavorInfo.${f}.nativeBuildInputs or []) enabledFlavors;
-  buildInputs = [ libgpgerror libassuan libcap libsecret ]
+  buildInputs = [ libgpgerror libassuan libcap libsecret texinfo]
```
Comment (jtojnar) [inline pkgs/tools/security/pinentry/default.nix:55]: Build tools should go to `nativeBuildInputs`.

File: pkgs/tools/security/pinentry/default.nix [hunk 4]
```diff
@@ -35,29 +36,29 @@ let
...
     ++ concatMap(f: flavorInfo.${f}.buildInputs or []) enabledFlavors;
 
   dontWrapGApps = true;
   dontWrapQtApps = true;
 
   patches = [
-    ./autoconf-ar.patch
```
Comment (jtojnar) [inline pkgs/tools/security/pinentry/default.nix:60]: You would also need to delete the file.

File: pkgs/tools/security/pinentry/default.nix [hunk 5]
```diff
@@ -66,6 +67,7 @@ pinentryMkDerivation rec {
   ];
 
   configureFlags = [
+    "--disable-doc" # disabling docs because pinentry doc build is broken
```
Comment (jtojnar) [inline pkgs/tools/security/pinentry/default.nix:70]: Could you share the error? Maybe we can fix it.

File: pkgs/top-level/aliases.nix [hunk 6]
```diff
@@ -427,6 +427,7 @@ mapAliases ({
   pinentry_qt = pinentry-qt; # added 2019-10-14
   pinentry_gnome = pinentry-gnome; # added 2019-10-14
   pinentry_qt5 = pinentry-qt; # added 2020-02-11
+  pinentry_efl = pinentry-efl; # added 2020-07-17
```
Comment (jtojnar) [inline pkgs/top-level/aliases.nix:430]: No need for alias when `pinentry_efl` was never used.

File: pkgs/tools/security/pinentry/default.nix (hunk 5 above)
Comment (bagnaram) [reply pkgs/tools/security/pinentry/default.nix]: The solution is to to this pinentry generation was that maintainer mode wasn't activated, leaving the `missing` script out of execution. https://autotools.io/automake/maintainer.html

**Comment by bagnaram** [comment]
Fixes addressed. Many thanks for the peer review

**Comment by bagnaram** [comment]
Any further action needed for this PR?

**Review by jtojnar — APPROVED** [review]
Changes look okay, do not see anything problematic in the upstream commits since 1.1.0: https://dev.gnupg.org/source/pinentry/history/master/

**Comment by ajs124** [comment]
This should be rebased, instead of having master merged into it.

**Comment by bagnaram** [comment]
Squashed and rebased to recent master. Please advise further action

File: pkgs/tools/security/pinentry/default.nix [hunk 7]
```diff
@@ -1,8 +1,9 @@
-{ fetchurl, mkDerivation, fetchpatch, stdenv, lib, pkgconfig, autoreconfHook, wrapGAppsHook
+{ fetchurl, texinfo, fetchgit, mkDerivation, fetchpatch, stdenv, lib, pkgconfig, autoreconfHook, wrapGAppsHook
 , libgpgerror, libassuan, qtbase, wrapQtAppsHook
 , ncurses, gtk2, gcr
+, callPackage, pkgs
```
Comment (SuperSandro2000) [inline pkgs/tools/security/pinentry/default.nix:4]: Please don't pass in pkgs.

File: pkgs/tools/security/pinentry/default.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/tools/security/pinentry/default.nix:56]: Can you clean this up?

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Can you please update the commit title to ``pinentry: 1.1.0 -> 2020-07-17``?

**Comment by bagnaram** [comment]
Thank  you for the comments. Updated repo with suggestions

File: pkgs/top-level/aliases.nix (hunk 6 above)
Comment (jtojnar) [reply pkgs/top-level/aliases.nix:461]: This does not seem to have been resolved.

**Comment by jtojnar** [comment]
And the branch commit history still looks messy/not squashed.

File: pkgs/tools/security/pinentry/default.nix [hunk 8]
```diff
@@ -35,20 +36,22 @@ let
     gnome3 = { bin = "gnome3"; flag = "gnome3"; buildInputs = [ gcr ]; nativeBuildInputs = [ wrapGAppsHook ]; };
     qt = { bin = "qt"; flag = "qt"; buildInputs = [ qtbase ]; nativeBuildInputs = [ wrapQtAppsHook ]; };
     emacs = { bin = "emacs"; flag = "emacs"; buildInputs = []; };
+    efl = { bin = "efl"; flag = "efl"; buildInputs = [ pkgs.enlightenment.efl ]; };
```
Comment (jtojnar) [inline pkgs/tools/security/pinentry/default.nix:39]: This does not evaluate.

File: pkgs/tools/security/pinentry/default.nix [hunk 9]
```diff
@@ -57,7 +60,6 @@ pinentryMkDerivation rec {
   dontWrapQtApps = true;
 
   patches = [
-    ./autoconf-ar.patch
```
Comment (jtojnar) [inline pkgs/tools/security/pinentry/default.nix:60]: Actually, without this patch, cross-compilation (`pkgsCross.aarch64-multiplatform.pinentry`) fails:

```
/nix/store/pmrhk324fkidrm5ffd5jckb21s9zys6r-bash-4.4-p23/bin/bash: ar: command not found
```

cc @nspin who added it in https://github.com/NixOS/nixpkgs/pull/68889

**Comment by fpletz** [comment]
Superseeded by #143120.

**Pull request closed without merging**